synapse/lib/view.py

```diff
@@ -138,12 +119,278 @@ async def __anit__(self, core, node):
+            await asyncio.sleep(0)
+            yield s_msgpack.un(byts)
+
+    async def getMerges(self):
+        '''
+        Yield the historical merges into this view.
+        '''
+        for lkey, bidn in self.core.slab.scanByPrefBack(self.bidn + b'hist:merge:time', db='view:meta'):
+            byts = self.core.slab.get(self.bidn + b'hist:merge:iden' + bidn, db='view:meta')
+            if byts is not None:
+                yield s_msgpack.un(byts)
+            await asyncio.sleep(0)
+
+    async def tryToMerge(self, tick):
+        # NOTE: must be called from within a nexus handler!
+
+        if self.merging: # pragma: no cover
+            return
+
+        if not await self.isMergeReady():
+            return
+
+        self.merging = True
+        await self.info.set('merging', True)
+
+        layr = self.layers[0]
+
+        layr.readonly = True
+        await layr.layrinfo.set('readonly', True)
+
+        merge = self.getMergeRequest()
+        merge['votes'] = [vote async for vote in self.getMergeVotes()]
+        merge['merged'] = tick
+
+        tick = s_common.int64en(tick)
+        bidn = s_common.uhex(merge.get('iden'))
+
+        lkey = self.parent.bidn + b'hist:merge:iden' + bidn
+        self.core.slab.put(lkey, s_msgpack.en(merge), db='view:meta')
+
+        lkey = self.parent.bidn + b'hist:merge:time' + tick + bidn
+        self.core.slab.put(lkey, bidn, db='view:meta')
+
+        await self.core.feedBeholder('view:merge:init', {'view': self.iden})
+
+        await self.initMergeTask()
+
+    async def setMergeVote(self, vote):
+        self.reqParentQuorum()
+        vote['created'] = s_common.now()
+        vote['offset'] = await self.layers[0].getEditIndx()
+        return await self._push('merge:vote:set', vote)
+
+    @s_nexus.Pusher.onPush('merge:vote:set')
+    async def _setMergeVote(self, vote):
+
+        self.reqParentQuorum()
+        s_schemas.reqValidVote(vote)
+
+        uidn = s_common.uhex(vote.get('user'))
+
+        self.core.slab.put(self.bidn + b'merge:vote' + uidn, s_msgpack.en(vote), db='view:meta')
+
+        tick = vote.get('created')
+        await self.tryToMerge(tick)
+
+        return vote
+
+    async def delMergeVote(self, useriden):
+        return await self._push('merge:vote:del', useriden, s_common.now())
+
+    @s_nexus.Pusher.onPush('merge:vote:del')
+    async def _delMergeVote(self, useriden, tick):
+
+        self.reqParentQuorum()
+        uidn = s_common.uhex(useriden)
+
+        byts = self.core.slab.pop(self.bidn + b'merge:vote' + uidn, db='view:meta')
+
+        await self.tryToMerge(tick)
+
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    async def initMergeTask(self):
+
+        if not self.merging:
+            return
+
+        if not await self.core.isCellActive():
+            return
+
+        self.mergetask = self.core.schedCoro(self.runViewMerge())
+
+    async def finiMergeTask(self):
+        if self.mergetask is not None:
```

Comment:
Testing this would require a graceful promotion in the middle of a merge task. Not impossible to setup but something we do need to test.